src/Interpreters/PreparedSets.cpp

```diff
@@ -43,6 +46,26 @@ bool PreparedSetKey::operator==(const PreparedSetKey & other) const
     return true;
 }
 
+String PreparedSetKey::toString() const
```

Comment:
Btw, in analyzer, we need to hash the same set differently depending on the type of left operand, check out the comment at `PlannerContext::createSetKey` in https://github.com/ClickHouse/ClickHouse/pull/48754

This solution solves the issue (at least fixes test `00700_decimal_compare`, but I'm uncertain if it's not too ad-hoc).